File: 20-valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const matchingCloseParentheses = {
        "{": "}",
        "[": "]",
        "(": ")"
    };

    const stack = [];

    for (const char of s) {
        if (char in matchingCloseParentheses) {
            stack.push(char)
        } else {
            if (!(stack.length > 0)) return false
            const openP = stack.pop();
            if (char !== matchingCloseParentheses[openP]) {
                return false
            }
        }
    }
    return stack.length === 0
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        matchingCloseParentheses = {
            "(": ")",
            "{": "}",
            "[": "]"
        }

        stack = []

        for char in s: 
            if char in matchingCloseParentheses:
                stack.append(char)
            else:
                if not stack: return False
                openP = stack.pop()
                if char != matchingCloseParentheses[openP]:
                    return False

        return len(stack) == 0
